### DISCUSSION TITLE
Stockfish exclusion of candidate moves explanation

### DISCUSSION THREAD

**Discussion by dlbbld** [official-stockfish/Stockfish#3913] [Q&A]
The following chess composition from Mark Liburkin is under https://miguel-ambrona.github.io/, INTERESTS.

It is a rare example of one side clearly winning, but Stockfish only sees it late. White in an apparent worse position is winning, for it most startling can trap the black queen.

The winning line starts with 1. Bd2+ b4 2. Bxb4+ Kb5 3. Nd6+ Kb6 4. Ba5+, Black's moves being forced. Stockfish does not see the line is winning until after 4. Ba5+ is on the board:

After 1. Bd2+ b4 2. Bxb4+ Kb5 3. Nd6+ Kb6 value for cp is -96, Stockfish sees no clear win:
 ```shell
position fen 2N5/p7/q1p2p2/kp6/3P4/PP2K3/5P2/2B5 w - - 0 1 moves c1d2 b5b4 d2b4 a5b5 c8d6 b5b6
go depth 30
info depth 30 seldepth 42 multipv 1 score cp -96 nodes 24942100 nps 1601110 hashfull 1000 tbhits 0 time 15578 pv b4c5 b6c7 d6e8 c7b7 e8f6 a6f1 f2f3 f1c1 e3e4 c1g5 f6g4 g5d5 e4f4 d5f7 f4e4 f7b3 g4e5 b3e6 e4f4 e6f6 f4g4 b7c8 f3f4 c8b7 f4f5 a7a5 a3a4 f6h6 e5g6 b7c7 g6e5 c7c8 e5c4 h6g7 g4f4 g7f6 f4g4 c8c7 c4e5
bestmove b4c5 ponder b6c7
```

After 4. Ba5+ is on the board, cp is -930, now only Stockfish sees the clear win for White:
 ```shell
position fen 2N5/p7/q1p2p2/kp6/3P4/PP2K3/5P2/2B5 w - - 0 1 moves c1d2 b5b4 d2b4 a5b5 c8d6 b5b6 b4a5
go depth 30
info depth 30 seldepth 44 multipv 1 score cp -930 nodes 70973400 nps 2592446 hashfull 971 tbhits 0 time 27377 pv b6a5 d6c4 a5b5 e3f4 a6a4 c4d6 b5a5 d6b7 a5b6 b3a4 b6b7 f4f5 b7c7 f5e6 f6f5 f2f4 c7d8 e6f5 d8e7 f5e5 e7d7 f4f5 d7e7 f5f6 e7f7 a4a5 f7g6 e5e6 g6g5 f6f7 g5g4 a3a4 g4g5 a5a6 g5g6
bestmove b6a5 ponder d6c4
```

Can someone please share some insight about Stockfish behaviour for this example?

**Comment by vondele** [comment]
I think it is not unexpected, has to sacrifice material, and the queen capture is only several moves later. The line is just very heavily pruned. Eventually it is found:
```
info depth 71 seldepth 95 multipv 1 score cp 4830 nodes 39089040673 nps 2635634 hashfull 1000 tbhits 0 time 14830977 pv b4a5 b6a5 d6c4 a5b5 e3f4 f6f5 f4g5 f5f4 f2f3 a6a4 c4d6 b5b6 b3a4 b6c7 d6c4 c6c5 d4c5 c7c6 g5f4 c6c5 f4f5 c5c4 f5e6 c4c3 f3f4 c3b3 f4f5 b3a3 f5f6 a3a4 f6f7 a4a5 e6e5 a5b5 f7f8q b5c6 f8a8 c6b6 a8b8 b6a6
```

**Reply by dlbbld** [reply]
Thanks, but I completely fail to understand. Stockfish does not see the move is winning before the move with depth 30 but afterwards with depth 30. So why does Stockfish not see the move is winning before playing with depth 31?

The mainline from the beginning is 1. Bd2+ b4 (forced) 2. Bxb4+ Kb5 (forced) 3. Nd6+ Kb6 (forced) 4. Ba5+ Kxa5 5. Nc4+ Kb5 6. Kf4 c5  7. d5 f5 8. Kg5 f4 9. f3. So before 4. Ba5+ there are 11 half-moves to get to a simple winning position.

So even more, if Ba5+ is on the board, Stockfish already sees with depth 10 the move is winning:
 ```
position fen 2N5/p7/q1p2p2/kp6/3P4/PP2K3/5P2/2B5 w - - 0 1 moves c1d2 b5b4 d2b4 a5b5 c8d6 b5b6 b4a5	  
go depth 10
info depth 10 seldepth 16 multipv 1 score cp 438 nodes 12532 nps 895142 tbhits 0 time 14 pv b6a5 d6c4 a5b5 e3f4 f6f5 f4e5 c6c5 d4d5 a6g6 c4d6 b5b6 d6f5
bestmove b6a5 ponder d6c4
 ```

But if I run Stockfish (restarting to avoid caching above line) with depth 11 before the move, it does not see it. Why?
 ```
position fen 2N5/p7/q1p2p2/kp6/3P4/PP2K3/5P2/2B5 w - - 0 1 moves c1d2 b5b4 d2b4 a5b5 c8d6 b5b6
go depth 11 searchmoves b4a5
info depth 11 seldepth 20 multipv 1 score cp -228 nodes 22065 nps 54080 tbhits 0 time 408 pv b4a5 b6a5 d6c4 a5b5 e3f4 c6c5 d4d5 f6f5 f4g5
bestmove b4a5 ponder b6a5
 ```

Stockfish sees the move is winning with depth 15:
 ```
position fen 2N5/p7/q1p2p2/kp6/3P4/PP2K3/5P2/2B5 w - - 0 1 moves c1d2 b5b4 d2b4 a5b5 c8d6 b5b6
go depth 15 searchmoves b4a5
info depth 15 seldepth 22 multipv 1 score cp 369 nodes 51340 nps 136180 tbhits 0 time 377 pv b4a5 b6a5 d6c4 a5b5 e3f4 f6f5 f4g5 f5f4 f2f3 a6a3 c4a3 b5b4 a3c4 b4c3 g5f4 c3d4 f4f5 d4c3 f3f4 a7a6
bestmove b4a5 ponder b6a5
 ```

So per my understanding, Stockfish does not evaluate Ba5+ beyond depth 15. But what is the reason Stockfish does not dive deeper into this line?

Three questions, many wrong assumptions, I guess. So I would be happy if you could clarify some of them.

**Reply by dlbbld** [reply]
In a nutshell, my question is if someone can describe with natural language the Stockfish reasoning not to evaluate the winning line at more depth.

**Reply by pb00067** [reply]
Because sacrificing material is loosing 99% of times in chess and SF cannot know in advance that sacrificing the bishop is winning.
Analyzing all moves with equal depth would decrease SF's strength dramatically.
So for not loosing to much time in analyzing moves which seem not promising (weak) the engine is heavily pruning them (=analyzing that moves with reduced depth). This has the natural consequence that good sacrifices moves are sometimes discovered by engines only at higher depths.

**Comment by vibe-ken** [comment]
分享快乐每一天... ...

**Comment by pb00068** [comment]
I reviewed this position and come to the conclusion that the main problem is null-move-pruning at following zugzwang position:
8/p7/q7/1kpP2K1/2N2p2/PP3P2/8/8 b - - 0 9 
![immagine](https://github.com/user-attachments/assets/e1e5d0c3-edd1-45ab-a3c6-60bbe0a6e30c)
Here black is forced to move the Queen and loosing it (either by direct recapture of royal fork on Nd6+),
but null-move-pruning let's the engine belief he can rest his Queen on a6 without moving at all.
In fact by disabling null-move-pruning with the black queen on A6, current SF understands that white is winning after ca. 1 minute:
![immagine](https://github.com/user-attachments/assets/120c6fcb-5049-4045-b207-872042ce0f8e)
